feat: new library package autoware control toolbox

boyali opened the pull request:
## Description

<!-- Write a brief description of this PR. -->

The ACT contains some essential control components in designing Autoware controllers and control
system-related packages and algorithms. The current release incorporates transfer function, state-space, and spline
modules.

## Related links

<!-- Write the links related to this PR. Private links should be clearly marked as private, for example, '[FOO COMPANY INTERNAL LINK](https://example.com)'. -->

## Tests performed

<!-- Describe how you have tested this PR. -->

## Notes for reviewers

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

yukkysaito commented:
It seems to be included change of unrelated files.
@TakaHoribe @takayuki5168  At first, can I ask you for a review in your team?

TakaHoribe commented:
@boyali Thank you. Please check the modified files in [this page](https://github.com/autowarefoundation/autoware.universe/pull/1971/files) and remove unneccesary changes.

boyali commented:
> @boyali Thank you. Please check the modified files in [this page](https://github.com/autowarefoundation/autoware.universe/pull/1971/files) and remove unneccesary changes.

Thank you for the quick response. I removed the unrelated files. There is only one in the .github directory (should I remove it as well)?

TakaHoribe commented:
>There is only one in the .github directory (should I remove it as well)?
Yes, please remove the .gitignore file as well.

TakaHoribe commented:
@boyali There already is the [interpolate/spline package](https://github.com/autowarefoundation/autoware.universe/tree/main/common/interpolation) in Autoware. Is there any difference?

TakaHoribe commented:
And maybe .github folder is not needed too.

boyali commented:
> 

Yes, this is the Eigen library version. The existing one only accepts std:: vector. There are also some additional implementations, such as binary tree search and single-point interpolation, linear interpolation.

boyali commented:
> And maybe .github folder is not needed too.

I think the cloud side created it automatically. But I check

takayuki5168 commented on control/autoware_control_toolbox/include/splines/bspline_interpolator_templated.hpp:
```diff
@@ -0,0 +1,730 @@
+// Copyright 2018-2022 The Autoware Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
As Horibe-san said, we have interpolation package.
What do you think about moving this PR's spline implementation to interpolation package.

takayuki5168 replied:
I'm not familiar with your project, but I got it.
I'll leave how to deal with this PR to @TakaHoribe !

takayuki5168 commented on control/autoware_control_toolbox/CMakeLists.txt:
```diff
@@ -0,0 +1,176 @@
+cmake_minimum_required(VERSION 3.10)
```
Let me ask you first whether there is any package in autoware.universe which will depend on the autoware_control_toolbox package after this PR is merged.
Will your nonlinear mpc depend on this package?

boyali replied:
Yes, but not only nonlinear MPC; we have delay compensation, adaptive controllers, and disturbance generator packages in the row. This ACT also makes designing controller, observer, and low-level control tasks easier.

boyali commented on control/autoware_control_toolbox/uncrustify.cfg:
```diff
@@ -0,0 +1,2807 @@
+# Uncrustify-0.69.0_f
```
Removed.

maxime-clem commented on control/autoware_control_toolbox/version.txt:
```diff
@@ -0,0 +1 @@
+v0.0.1
```
I have not seen this file in other packages. Is it necessary ?
I think the version of ROS packages is supposed to be maintained in the `package.xml` file.

boyali replied:
Sorry, it was for my reference; I remove it.

maxime-clem commented on control/autoware_control_toolbox/include/splines/bsplines_smoother.hpp:
```diff
@@ -0,0 +1,124 @@
+  // the same size.
+  void InterpolateInCoordinates(
+    Eigen::MatrixXd const &ybase,
+    Eigen::MatrixXd &data_tobe_interpolated);
```
This function should be made `const`.

maxime-clem commented on control/autoware_control_toolbox/documentation/SplineModules.md:
```diff
@@ -0,0 +1,160 @@
+\mathbf{y}
+$$
+
+where on $\hat{y}$ represents the interpolated and smoothed signal, whereas $y$ on the left-hand side is the original
```
```suggestion
where on $\hat{y}$ represents the interpolated and smoothed signal, whereas $y$ on the right-hand side is the original
```

maxime-clem commented on control/autoware_control_toolbox/include/utils_act/act_utils.hpp:
```diff
@@ -0,0 +1,621 @@
+}
+
+template<typename T, typename std::enable_if_t<std::is_floating_point_v<T>> * = nullptr>
+constexpr T angleDistance(T const &theta, T const &theta_ref)
```
This can probably be replaced by https://github.com/autowarefoundation/autoware.universe/blob/main/common/motion_common/src/motion_common/trajectory_common.cpp#L36-L39

maxime-clem commented on control/autoware_control_toolbox/include/utils_act/act_utils.hpp:
```diff
@@ -0,0 +1,621 @@
+}
+
+template<typename T, typename std::enable_if_t<std::is_floating_point_v<T>> * = nullptr>
+T wrapToPi(T const &angle)
```
Can this be replaced with the following ?
https://github.com/autowarefoundation/autoware.universe/blob/34c9947da31dec92804c27375714d7380e28614f/common/autoware_auto_common/include/helper_functions/angle_utils.hpp#L50

boyali replied:
I need to check it for all angle difference possibilities first in the Python than with the NMPC, tomorrow, I will do it and let you know. If there is no incompatibility, I will remove it. Thank you for your valuable reviews.

maxime-clem commented on control/autoware_control_toolbox/CMakeLists.txt:
```diff
@@ -0,0 +1,164 @@
+autoware_package()
+
+# -------------  FLAGS ---------------------------------------------------
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(
+    -Wall
+    -Wextra
+    -Wpedantic
+    # -Wuseless-cast
+    # -Wno-unused-parameter
+    # -Wno-unused-variable
+    # -Wno-narrowing
+    # -Wno-error=unused-but-set-variable
+    -Wno-conversion
+    -Wno-error=sign-conversion
+    -Wno-sign-compare
+    # -Wdouble-promotion
+    # -Wfloat-equal
+    -Waddress
+    -fvisibility=default
+    -fPIC) # visibility hidden or default is set here.
+endif()
```
Custom flags should be removed.

boyali replied:
I put these flags because there are codes I can't change. For example, if I remove it, it gives an error for simple warnings. For example 

```
ent signedness: ‘Eigen::Index’ {aka ‘long int’} and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   64 |   for (Eigen::Index k = 0; k < npoints_; k++)
```
I am using Eigen::Index and std::vectors at the same time; for a simple loop, I need to introduce unnecessary static cast back and forth. 

If we can keep it, it would be appreciated.

boyali replied:
For this case, I fixed it. But the same sign comparison error will hit in the NMPC.

maxime-clem commented on control/autoware_control_toolbox/CMakeLists.txt:
```diff
@@ -0,0 +1,164 @@
+#ament package generation and installing
+ament_auto_package(INSTALL_TO_SHARE)
+
+#-------------------------  PROTOTYPES __________________________________________
+## CONTROL
+ament_auto_add_executable(main_tf usage_examples/control/main_tf.cpp)
+ament_auto_add_executable(main_pade usage_examples/control/main_pade.cpp)
+ament_auto_add_executable(main_tf2ss usage_examples/control/main_tf2ss.cpp)
+ament_auto_add_executable(main_tf_algebra usage_examples/control/main_tf_algebra.cpp)
+ament_auto_add_executable(main_balance usage_examples/control/main_balance.cpp)
+ament_auto_add_executable(main_signal_generator usage_examples/control/main_signal_generator.cpp)
+
+##---------------------- SPLINEs ------------------------------------------------
+
+add_executable(main_spline_pcg usage_examples/splines/main_spline_eigen_pcg.cpp)
+target_link_libraries(main_spline_pcg ${CONTROL_LIB} fmt::fmt stdc++fs)
+
+add_executable(main_bspline_smoother usage_examples/splines/main_bspline_smoother.cpp)
+target_link_libraries(main_bspline_smoother ${CONTROL_LIB} fmt::fmt stdc++fs)
+
+add_executable(main_bspline_interp usage_examples/splines/main_bspline_interpolator.cpp)
+target_link_libraries(main_bspline_interp ${CONTROL_LIB} fmt::fmt stdc++fs)
+
+add_executable(main_bspline_interp_template usage_examples/splines/main_bspline_interpolator_templated.cpp)
+target_link_libraries(main_bspline_interp_template ${CONTROL_LIB} fmt::fmt stdc++fs)
+
+add_executable(main_unwrap usage_examples/helper_functions/main_unwrap.cpp)
+target_link_libraries(main_unwrap ${CONTROL_LIB} fmt::fmt stdc++fs)
+
+add_executable(main_1d_interp usage_examples/helper_functions/main_linear_1dinterp.cpp)
+target_link_libraries(main_1d_interp ${CONTROL_LIB} fmt::fmt stdc++fs)
+
+add_executable(main_diff_matt usage_examples/helper_functions/main_difference_matrix.cpp)
+target_link_libraries(main_diff_matt ${CONTROL_LIB} fmt::fmt stdc++fs)
```
These are nice but they may increase the build time of the package but are not necessary.
Can you make their build conditional on some CMake flag ?

boyali replied:
Commented  them out (disabled their compilation).

TakaHoribe commented:
@boyali Please fix the CI. Required items need to pass to merge.

boyali commented:
> @boyali Please fix the CI. Required items need to pass to merge.

I am fixing it.

boyali commented:
@TakaHoribe -san,
I fixed many CI checks. There are problems with CodeScene but the problems it complains about are right (some of them) and most of them are related to the cpp examples I put (long main function).

It complains about the similar structure of the methods but they are different methods. 

`The module contains 2 functions with similar structure: ns_splines::BSplineInterpolator::getFirstDerivative,ns_splines::BSplineInterpolator::getSecondDerivative` or

`The module contains 3 functions with similar structure: TEST:ACTcontrol:multiplyTFfactors,TEST:ACTcontrol:subtractTFfactor,TEST:ACTcontrol:sumTFfactor`.

Should I fix all of them?

boyali commented:
Fixed all except variable name spell checks.

boyali commented:
@TakaHoribe -san, 
I addressed all your concerns, removed duplicated methods and modules, and tested it with the NMPC. I have to recommend some modifications to the spline modules and asked for PR for your consideration. 
https://github.com/autowarefoundation/autoware.universe/pull/2202

TakaHoribe commented:
@boyali This PR still has spline-related code (readme/directory). Please check the changed-file tab and remove them.

boyali commented:
> @boyali This PR still has spline-related code (readme/directory). Please check the changed-file tab and remove them.

Thank you for the review. I removed natural spline description from the readme files.
